### PR TITLE
fix: prefetch in RSC

### PR DESCRIPTION
**Pull request by juliusmarminge**
Fixes 2 bugs:

- `_def` sometimes being undefined. Fixed by adding a flat proxy and making the `_def` a synchronous getter.
image

- race condition when queries are fast and resolves before hitting the network boundary resulting in serialized data being put in the query cache (upstream fix https://github.com/TanStack/query/pull/7615/files)
image

## Fixed result:

https://github.com/trpc/trpc/assets/51714798/ac9ce585-ba05-4f6a-89d5-7ea42a44290b

### CODE REVIEW

File: patches/@tanstack__query-core@5.45.0.patch
Comment (juliusmarminge) [inline patches/@tanstack__query-core@5.45.0.patch:1]: https://github.com/TanStack/query/pull/7615

File: packages/server/src/unstable-core-do-not-import/router.ts [hunk 1]
```diff
@@ -276,40 +276,41 @@ export function createCallerFactory<TRoot extends AnyRootTypes>() {
...
+            const fullPath = [path, ...opts.path].join('.');
+
+            const procedure = _def.procedures[fullPath] as AnyProcedure;
+
+            let ctx: Context | undefined = undefined;
+            try {
+              return Promise.resolve(
+                isFunction(ctxOrCallback) ? ctxOrCallback() : ctxOrCallback,
+              ).then((resolvedContext) => {
+                ctx = resolvedContext;
```
Comment (juliusmarminge) [inline packages/server/src/unstable-core-do-not-import/router.ts:293]: i don't think this is actually necessary anymore

**Review by KATT — APPROVED** [review]

File: packages/react-query/package.json [hunk 2]
```diff
@@ -64,14 +64,14 @@
   },
   "dependencies": {},
   "peerDependencies": {
-    "@tanstack/react-query": "^5.40.0",
+    "@tanstack/react-query": "^5.49.2",
```
Comment (KATT) [inline packages/react-query/package.json:67]: nice

Comment (juliusmarminge) [reply packages/react-query/package.json:67]: fuck I hate that manypkg bumps peerDep range. We can keep it at 5.40 assuming most ppl wont use this RSC void prefetch thing?

Comment (KATT) [reply packages/react-query/package.json:67]: i think we should bump it and i'm fine with bumping this with any minor/patch release

**Pull request merged**